src/hooks/use-poll.js

```diff
@@ -1,20 +1,58 @@
+import { useDataQuery } from '@dhis2/app-runtime'
 import { useRef, useEffect } from 'react'
 
-export const usePoll = (fn, pollInterval) => {
+const useConst = factory => {
+    const ref = useRef(null)
+    if (ref.current === null) {
+        ref.current = factory()
+    }
+    return ref.current
+}
+
+const useLazyInterval = (fn, interval) => {
     const intervalId = useRef(null)
 
-    const start = (...args) => {
-        intervalId.current = setInterval(() => fn(...args), pollInterval)
-    }
     const cancel = () => {
         if (intervalId.current) {
             clearInterval(intervalId.current)
         }
     }
+    const start = (...args) => {
+        cancel()
+        intervalId.current = setInterval(() => fn(...args), interval)
+    }
 
     useEffect(() => {
         return cancel
     }, [])
 
     return { start, cancel }
 }
+
+export const usePoll = ({ query, interval, checkDone }) => {
+    const wrappedQuery = useConst(() => ({
+        poll: query,
+    }))
+    const { refetch, error, data } = useDataQuery(wrappedQuery, {
+        lazy: true,
+        onComplete: data => {
+            if (checkDone(data.poll)) {
+                cancel()
+            }
+        },
+    })
+    const { start, cancel } = useLazyInterval(refetch, interval)
+
+    useEffect(() => {
+        if (error) {
+            cancel()
+        }
+    }, [error])
+
+    return {
+        start,
+        cancel,
+        error,
+        data: data?.poll,
+    }
+}
```

Comment:
Instead of accepting a query, I think it makes a lot more sense to pass the refetch function to this hook. Then the query can be passed statically to the `useDataQuery` hook in the respective files (plus less "magic" and more explicit code)

Reply:
That's how I did it at one point https://github.com/dhis2/data-administration-app/pull/630/commits/ba78150ab830426397d8f5c5c1346bd1acbb9962 but I wasn't too happy with the fact that each poll hook needed to have a `useEffect` doing effectively the same thing. Do you think that approach was fine and I should revert that commit or is there a kind of middle-ground with the best ob both worlds?

Reply:
What I meant is something like this:

```js
// use-poll.js
export const usePoll = ({ refetch, error, interval, checkDone }) => {
    const { start, cancel } = useLazyInterval(refetch, interval)

    useEffect(() => {
        if (error) {
            cancel()
        }
    }, [error])

    return {  start, cancel }
}
```

```js
// use-analytics-poll.js
import { usePoll } from '../../hooks/use-poll'

const pollQuery = {
    analyticsTable: {
        resource: 'system/tasks/ANALYTICS_TABLE',
        id: ({ jobId }) => jobId,
    }
}

export const useAnalyticsPoll = () => {
    const hookReturnValue = useDataQuery(pollQuery)
    const { start, cancel } = usePoll({
        ...hookReturnValue,
        interval: 3000,
        checkDone: data => data.some(t => t.completed),
    })

    return { ...hookReturnValue, start, cancel }
}
```

Reply:
Cool I'll do it that way :+1: